### PR TITLE
improve logging, separate display/debug logs

### PR DESCRIPTION
**Pull request by whilei**
### problem
Existing logging isn't as friendly-looking as it could be. The logs are purely event/protocol derived and include a lot of often-noisy information.

### solution
Separate "debug"-y logs from "displayable" logs; _ie_ end-user-facing vs. developer-facing. Base display logs not only on events (via `mux` subscriptions), but also on time-interval-based status logging.

The proposed solution:
- modifies __display log line formatting__ toward simplicity, consistency, and grouping related ideas (like block import statistics _processed_, _ignored_, _queued_, etc.). It also introduces thematic coloring, icon-ification for modes, events, and tallies.
- **collects all existing** `glog.V` **protocol/event logs to files** in `&lt;chaindir&gt;/log` with default verbosity `Debug`. These files are then **rotated automatically** based on size and age.
- introduces `glog.D` for logs which should only be rendered to stderr; these **display logs** are also verbosity-aware, and can be adjusted. Current default is `--display=2` on a practical scale of `0(totally silent)-3(include event subscriptions)`, where `2` prints mostly only interval-based status logs (ala `--log-status`), like:
```
I1203 07:10:53 SYNC Import   ▶︎⟪#4950584 0xfca0d22…⟫ +⟪   4 blks   40 txs   893225 mgas⟫ ✌︎︎︎⟪12/25 peers⟫ 🂓🁥🁦🂓
```
while `--display=3` includes subscription-based logs like
```
I1203 07:16:04 ◼⋯⋯◼ Insert blocks=⟪processed=   1 queued=   0 ignored=   0 txs=   0⟫ ◼=⟪n= 4950602 hash=0xb044b6d… time=54.083401725s ago⟫ took=⟪5.868158ms⟫
```
- establishes default verbosity+severity thresholds **enabling any existing** `severity=Error` logs from `glog.V` **to _also_ be written to stderr**, along with their file trace
- introduces `--neckbeard` flag for developers who'd rather get all normal `glog.V` debug-y logs to stderr and don't like pretty colors

### discussion
❓ Do we like the colors? Any suggestions, feedback, or wishes for the proposed formatting?
❓ Should we move default `--display` verbosity to `3` to also include subscription-based logs?

### todo
- [x] include miner display logs
- [x] fix tests
- [x] write tests
- [x] ensure symbols/icons work on all os's

🖌 There's a lot of commits and they're pretty messy (not always `problem/solution` format, etc...). Will squash before merge, but left as-is for sake of transparency and granularity for now.

### examples
ℹ️ Note in the screenshots the small terminal window below where `tail` is following the relevant debug log files (`~/Library/EthereumClassic/mainnet/log/geth.INFO`).

__Current default__: `--display=3`
screen shot 2017-12-03 at 07 24 29

`--display=2`
screen shot 2017-12-03 at 07 15 35

`--neckbeard`
screen shot 2017-12-03 at 07 25 53

### iconography
__Event logs__ have associated icons:
> __Downloader [Start|Done|Fail]__: `◼⋯⋯⬇` 
```
I1203 07:48:39 ◼⋯⋯⬇ Start ⟪Peer id=3355c6b44e100a99 eth/63 [Parity/v1.8.1-beta-0e4a06d-20171020/x86_64-linux-gnu/rustc1.21.0] [hs 0.00/s, bs 0.00/s, rs 0.00/s, ss 0.00/s, miss    0, rtt 20s]⟫ hash=⟪0x2d6e961…⟫ TD=⟪171941664579510193825⟫
```
> __BlockchainInsertBlocks__: `◼⋯⋯◼`
```
I1203 07:50:04 ◼⋯⋯◼ Insert blocks=⟪processed=   1 queued=   0 ignored=   0 txs=   7⟫ ◼=⟪n= 4950744 hash=0x1ffed1e… time=20.567166733s ago⟫ took=⟪27.439103ms⟫
```
> __BlockchainInsertForkedBlock__: `◼⋯⦦◼`
```
I1203 07:51:31 ◼⋯⦦◼ Insert forked block=⟪n= 4950749 hash=0x3776d82…⟫
```
> __HeaderchainInsertHeaders__: `◼⋯⋯❐`
```
I1203 07:58:11 ◼⋯⋯❐ Insert headers=⟪processed=2048 ignored=   0⟫ ❐=⟪n=8576 hash=0x0ddc763…⟫took=⟪433.059351ms⟫
```
> __MinedBlock__: `◼⋯⋯⟠`
```
I1206 08:25:17 ◼⋯⋯⟠ Mined ◼=⟪n=     139 hash=0xc08322b… coinbase=0x3d1cf2f… txs=  0 uncles=0⟫
```

__SYNC logs__ have associated icons:
> __Discover__: "spins" through `{"➫", "➬", "➭"}`
```
I1203 07:59:03 SYNC Discover ➫⟪#      0 0xd4e5674…⟫ ~⟪   0 blks    0 txs  0 mgas  /sec⟫ ✌︎︎︎⟪ 0/25 peers⟫
I1203 07:59:33 SYNC Discover ➬⟪#      0 0xd4e5674…⟫ ~⟪   0 blks    0 txs  0 mgas  /sec⟫ ✌︎︎︎⟪ 0/25 peers⟫
I1203 08:00:03 SYNC Discover ➭⟪#      0 0xd4e5674…⟫ ~⟪   0 blks    0 txs  0 mgas  /sec⟫ ✌︎︎︎⟪ 1/25 peers⟫
```
> __FullSync__: `◉`
```
I1203 08:02:33 SYNC FullSync ◉⟪#  33600 0x45b38c6…⟫ ~⟪  66 blks    0 txs  0 mgas  /sec⟫ ✌︎︎︎⟪ 4/25 peers⟫
```
> __FastSync__: `◎`
```
I1203 08:02:33 SYNC FastSync ◎⟪#  33600 0x45b38c6…⟫ ~⟪  66 blks    0 txs  0 mgas  /sec⟫ ✌︎︎︎⟪ 4/25 peers⟫ height=⟪4950785 0.68%⟫
```
> __Import__ (tracking at head): `▶`
```
I1203 07:53:31 SYNC Import   ▶︎⟪#4950761 0x6c9f16a…⟫ +⟪  18 blks  140 txs  3670871 mgas⟫ ✌︎︎︎⟪ 6/25 peers⟫ 🁼🁣🁦🂓🂌🁦🁣🁣🁣🂓🁥🁣🁴🁣🁴🁼🁣🁤
```
  > __"Domino Effect"__: the number of dominos relates the number of blocks imported in the given period, while the number of dots on each domino correspond to the number of txs in each relative block. When the total number of blocks imported is greater than 20, a single ellipsis is added to prevent clogging the screen with possibly hundreds of dominos.
----

Rel issues logrotation: #421, noisy log lines: #410, general improve status log: #127 
Rel milestone 4.2: https://github.com/ethereumproject/go-ethereum/milestone/12

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -684,9 +691,21 @@ func (self *BlockChain) LoadLastState(dryrun bool) error {
...
 	blockTd := self.GetTd(self.currentBlock.Hash())
 	fastTd := self.GetTd(self.currentFastBlock.Hash())
-	glog.V(logger.Info).Infof("Last header: #%d [%x…] TD=%v", self.hc.CurrentHeader().Number, self.hc.CurrentHeader().Hash().Bytes()[:4], headerTd)
-	glog.V(logger.Info).Infof("Last block: #%d [%x…] TD=%v", self.currentBlock.Number(), self.currentBlock.Hash().Bytes()[:4], blockTd)
-	glog.V(logger.Info).Infof("Fast block: #%d [%x…] TD=%v", self.currentFastBlock.Number(), self.currentFastBlock.Hash().Bytes()[:4], fastTd)
+	glog.V(logger.Warn).Infof("Last header: #%d [%x…] TD=%v", self.hc.CurrentHeader().Number, self.hc.CurrentHeader().Hash().Bytes()[:4], headerTd)
+	glog.V(logger.Warn).Infof("Last block: #%d [%x…] TD=%v", self.currentBlock.Number(), self.currentBlock.Hash().Bytes()[:4], blockTd)
+	glog.V(logger.Warn).Infof("Fast block: #%d [%x…] TD=%v", self.currentFastBlock.Number(), self.currentFastBlock.Hash().Bytes()[:4], fastTd)
+	glog.D(logger.Warn).Infof("Local head header:     #%s [%s…] TD=%s",
+		logger.ColorGreen(strconv.Itoa(int(self.hc.CurrentHeader().Number.Uint64()))),
```
[the remark this answers is not in the document]
Comment (whilei) [reply core/blockchain.go]: Note to self: still TODO.

**Comment by whilei** [comment]
Possibly controversial:  d1d37df

**Comment by whilei** [comment]
Also notable -- there's a lot of log configuration possibilities, and it might be nice to be able to see your settings once upfront... maybe overkill though. Lines like this would possibly be replaced/removed when we get file configuration

```
I1212 15:40:18 Debug log config: verbosity=5 log-dir=/Users/ia/Library/EthereumClassic/mainnet/log vmodule=*
I1212 15:40:18 Display log config: display=3 status=sync=33s
I1212 15:40:18 Machine log config: mlog=json mlog-dir=/Users/ia/Library/EthereumClassic/mainnet/mlogs
```

- [ ] add line for log file rotation settings

**Comment by splix** [comment]
If it supposed to fix #127 it doesn't look as requested output

**Comment by whilei** [comment]
Right -- I've added and modified a few features _for consideration_, keeping in mind that revising the formatting and defaults at this point can be done _simply_, since the critical infrastructure to support flexible display output as a separate concern from "debug" output (which will go to an auto-rotating and -compressing filesystem store) is in place now (with #426 and this PR).

I'm bringing your [latest comment from #127](> https://github.com/ethereumproject/go-ethereum/issues/127#issuecomment-346515027) over here to consolidate the conversation: 

> 1. just nothing, looks like it stuck. tried few times, blank screen, or error logs if try to emulate network problems. so geth works, but doesn't give status feedback log
> 3. there will be _only_ sync and import in that status. it's not logs
> 4. but it will always be 98% or 99%, except initial import, why? and if you need it, time in hours makes more sense during initial import, rather than percentage
> 5. ah, I see, it was on `STATUS SYNC Discover` mode. I didn't plan Discover mode initially, only sync, so didn't realize it's not a sync. let's discuss it on separate ticket do we need discover feedback and if we need how it should look like (because _current block num_ and _blks/txs/mgas sec_ are definitely meaningless in that context)
> 6. let's make good default output. users will never dig into command line options for that, they will quit instead

----

- Still not sure about `1`. Still happening for you from master/4.1.x? Can you show the command you're using to start geth? Neither @tzdybal or I have been able to reproduce on Linux or Mac machines.

- `3` -- what do you mean exactly? Remove the event logs (like Chain `Insert`) from defaults? 

- `4` - I like the percentage synced output, it's useful and familiar to me. If we add or switch to time in hours, do you mean a guess at how many hours are left in the sync? I expect that guess would be kind of inexact, since QoS and download rates can vary widely. Or do you mean how many hours geth has been running?

- `5` - I agree that we can remove `_current block num_ and _blks/txs/mgas sec_` from the `Discover` output, and perhaps replace it with Quality of Service? Or log connected/disconnected peers (though that would be very verbose). 

Notes to [above comment](https://github.com/ethereumproject/go-ethereum/pull/423#issuecomment-351739137):
- The originally spec'd log line format had ` 11/18/25 peers`, where `11`=downloadingFrom, `18`=connected, and `25`=max peers. Since our downloader is synchronous, geth only ever downloads from 1 peer, so I removed that `11/` part from the format since it would be redundant and non-informative
- block hex display was originally formatted as `d819…f71c`, which is still a possibility. For me it seems less clear/useful than the current proposed `d819eb4...` format because because `d819…f71c` is unclear to me if it's show the first and final hex characters of a single block, or if it's showing n-characters from two blocks.


- Definitely agreed that we want to make default display output as good as possible; if you have any more specific ideas or criticisms for any log lines in particular please feel free -- I hope the "Iconography" section above would make it easier to get a sense of the big picture and be able to talk specifically about each proposed line

**Comment by splix** [comment]
1. can't run it right now, as it fails to build `logger/glog/glog.go:1243: undefined: sort.Slice`
2. yes, remove insert
3. ok, leave percents if you think it's easier
4. just discover, it's enough information

I've updated original issue #127 , please take a look. That's exactly what we need, a goal. I suggest to reach main goal before adding extra features and extended outputs

**Comment by whilei** [comment]
What's your `go version`? Maybe [`sort.Slice`](https://golang.org/pkg/sort/#Slice) was implemented in a later version... currently geth requires 1.8.x and [we're about to bump to 1.9.2](https://github.com/ethereumproject/go-ethereum/pull/430)

**Comment by splix** [comment]
`go version go1.7.5 linux/amd64`, it seems to be latest version for debian `Debian 4.9.30-2+deb9u3 (2017-08-06) x86_64 GNU/Linux` :(

**Comment by whilei** [comment]
Will this work? https://debian.pkgs.org/sid/debian-main-amd64/golang-go_1.9~2_amd64.deb.html

**Comment by splix** [comment]
That's for Debian Sid, not released yet. I have Debian 9, latest stable, it has only go 1.7. But I'll think what I can do, there should be a way to install go 1.8 on debian 9

**Comment by tzdybal** [comment]
@splix, what about: https://packages.debian.org/stretch/golang-1.8?

**Comment by whilei** [comment]
just an update from today -
screen shot 2017-12-14 at 16 58 47

still todo is modify the date formatting

also note that i'm intentionally running at 30s intervals for developer sanity ;)

**Comment by splix** [comment]
looks good now! Also let's remove `I1214` (what it means?)

**Comment by tzdybal** [comment]
I for INFO, and rest is a date (month and day).

**Comment by whilei** [comment]
I also propose (re)considering the block number display formatting's use of the `#` symbol, as in `#5023182`. 

Maybe I'm getting too lexicographical but the [`#` symbol is also known simultaneously as the "hash" and "number" symbol](https://duckduckgo.com/?q=hash+symbol&atb=v79-7__&ia=web), which I think makes it kind of funky to use as one or the other, when blocks have both both numbers and hashes. 

An alternative might be use `⬡` or `⬣` as a reference to a hash "hex" instead, and `n` for "number"? Just throwing ideas out here, but I'd like to at least address the hash/number symbol ambiguity idea.

**Comment by tzdybal** [comment]
When CPU mining is enabled, there is absolutely no information about mining in default display. This is probably not very common case, but I think that some logs should be enabled.

File: cmd/geth/flags.go [hunk 2]
```diff
@@ -142,36 +142,69 @@ var (
...
 	DisplayFlag = cli.IntFlag{
 		Name:  "display",
-		Usage: "Display verbosity: 0=silent, 1=basics, 2=status, 3=events",
-		Value: 2,
+		Usage: "Display verbosity: 0=silent, 1=basics, 2=status, 3=status+events",
+		Value: glog.DefaultDisplay,
+	}
+	DisplayFormatFlag = cli.StringFlag{
+		Name:  "display-fmt",
+		Usage: "Display format",
```
Comment (tzdybal) [inline cmd/geth/flags.go]: Please add information about possible values (`default`, `dash` and `green`).

**Comment by whilei** [comment]
@tzdybal I agree. You, @splix? 

The log event is ready for inclusion, we just need to spec the line... how about

```
2017-02-03 16:44:00  Discover                                                              0/25 peers
2017-02-03 16:45:00  Discover                                                              1/25 peers
2017-02-03 16:46:00  Fast   #2481951 of #3124363    79.4%   1211/  554    blk/mgas sec     6/25 peers
2017-02-03 16:47:00  Fast   #2689911 of #3124363    86.1%    611/  981    blk/mgas sec     6/25 peers
2017-02-03 16:48:00  Fast   #2875913 of #3124363    92.0%    502/  760    blk/mgas sec     4/25 peers
2017-02-03 16:49:00  Sync   #3124227 of #3124363 c76c34e7   77/ 242/ 7 blk/tx/mgas sec     4/25 peers
2017-02-03 16:50:00  Sync   #3124247 of #3124363 75e48eff   51/  51/ 5 blk/tx/mgas sec     4/25 peers
2017-02-03 16:51:00  Sync   #3124567 of #3124363 9af334ae  117/ 129/11 blk/tx/mgas sec     5/25 peers
2017-02-03 16:52:00  Sync   #3124787 of #3124363 1e3a8351    9/   6/ 1 blk/tx/mgas sec     7/25 peers
2017-02-03 16:52:05  Import #3124788             84e11ff4        15/ 7 tx/mgas            10/25 peers
2017-02-03 16:52:25  Import #3124789             9e45a241         5/ 1 tx/mgas            12/25 peers
2017-02-03 16:52:45  Import #3124790             d819f71c         0/ 0 tx/mgas            18/25 peers
2017-02-03 16:52:46  Mined #3124791             b719f31b         7/ 1 tx/mgas            18/25 peers
```

**Comment by splix** [comment]
Looks great!
Can we align all units to (by `tx/mgas`):
```
2017-02-03 16:44:00  Discover                                                              0/25 peers
2017-02-03 16:45:00  Discover                                                              1/25 peers
2017-02-03 16:46:00  Fast   #2481951 of #3124363    79.4%   1211/  554    blk/mgas sec     6/25 peers
2017-02-03 16:47:00  Fast   #2689911 of #3124363    86.1%    611/  981    blk/mgas sec     6/25 peers
2017-02-03 16:48:00  Fast   #2875913 of #3124363    92.0%    502/  760    blk/mgas sec     4/25 peers
2017-02-03 16:49:00  Sync   #3124227 of #3124363 c76c34e7   77/ 242/ 7 blk/tx/mgas sec     4/25 peers
2017-02-03 16:50:00  Sync   #3124247 of #3124363 75e48eff   51/  51/ 5 blk/tx/mgas sec     4/25 peers
2017-02-03 16:51:00  Sync   #3124567 of #3124363 9af334ae  117/ 129/11 blk/tx/mgas sec     5/25 peers
2017-02-03 16:52:00  Sync   #3124787 of #3124363 1e3a8351    9/   6/ 1 blk/tx/mgas sec     7/25 peers
2017-02-03 16:52:05  Import #3124788             84e11ff4        15/ 7     tx/mgas        10/25 peers
2017-02-03 16:52:25  Import #3124789             9e45a241         5/ 1     tx/mgas        12/25 peers
2017-02-03 16:52:45  Import #3124790             d819f71c         0/ 0     tx/mgas        18/25 peers
2017-02-03 16:52:46  Mined  #3124791             b719f31b         7/ 1     tx/mgas        18/25 peers
```

or to (left):
```
2017-02-03 16:44:00  Discover                                                              0/25 peers
2017-02-03 16:45:00  Discover                                                              1/25 peers
2017-02-03 16:46:00  Fast   #2481951 of #3124363    79.4%   1211/  554 blk/mgas sec        6/25 peers
2017-02-03 16:47:00  Fast   #2689911 of #3124363    86.1%    611/  981 blk/mgas sec        6/25 peers
2017-02-03 16:48:00  Fast   #2875913 of #3124363    92.0%    502/  760 blk/mgas sec        4/25 peers
2017-02-03 16:49:00  Sync   #3124227 of #3124363 c76c34e7   77/ 242/ 7 blk/tx/mgas sec     4/25 peers
2017-02-03 16:50:00  Sync   #3124247 of #3124363 75e48eff   51/  51/ 5 blk/tx/mgas sec     4/25 peers
2017-02-03 16:51:00  Sync   #3124567 of #3124363 9af334ae  117/ 129/11 blk/tx/mgas sec     5/25 peers
2017-02-03 16:52:00  Sync   #3124787 of #3124363 1e3a8351    9/   6/ 1 blk/tx/mgas sec     7/25 peers
2017-02-03 16:52:05  Import #3124788             84e11ff4        15/ 7 tx/mgas            10/25 peers
2017-02-03 16:52:25  Import #3124789             9e45a241         5/ 1 tx/mgas            12/25 peers
2017-02-03 16:52:45  Import #3124790             d819f71c         0/ 0 tx/mgas            18/25 peers
2017-02-03 16:52:46  Mined  #3124791             b719f31b         7/ 1 tx/mgas            18/25 peers
```

it seems that second option is more logical and more flexible for extensions

**Comment by whilei** [comment]
screen shot 2017-12-19 at 13 48 58

File: cmd/geth/log_display_dash.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+import (
+	"time"
+	"gopkg.in/urfave/cli.v1"
+	"github.com/ethereumproject/go-ethereum/eth"
+	"github.com/gizak/termui"
+	"github.com/ethereumproject/go-ethereum/logger/glog"
+	"github.com/ethereumproject/go-ethereum/logger"
+	"github.com/ethereumproject/go-ethereum/core"
+	"fmt"
+)
```
[the remark this answers is not in the document]
Comment (whilei) [reply cmd/geth/log_display_dash.go]: Done.

**Comment by whilei** [comment]
FYI -- waiting on rustup brokennes https://github.com/rust-lang-nursery/rustup.rs/issues/1316

**Comment by splix** [comment]
@whilei tested with last changes, looks really good now.

The only thing I've noticed is that it put duplicate lines sometimes:
```
2017-12-30 17:40:52 Import    #5112945              21096f96          1/ 0 txs/mgas         12/25 peers
2017-12-30 17:40:53 Import    #5112945              21096f96          6/ 0 txs/mgas         12/25 peers
...
2017-12-30 17:43:45 Import    #5112955              0a128794          8/ 0 txs/mgas         13/25 peers
2017-12-30 17:43:46 Import    #5112955              0a128794          5/ 0 txs/mgas         13/25 peers
...
2017-12-30 17:44:42 Import    #5112959              93acc208         18/ 0 txs/mgas         13/25 peers
2017-12-30 17:44:43 Import    #5112959              93acc208          0/ 0 txs/mgas         13/25 peers
```

It looks strange, especially the fact that block hash is same, but number of transactions is different.

**Comment by splix** [comment]
Also, when it imports few blocks at once, let's add number of additional blocks, like:
```
2017-12-30 22:11:42 Import    #5114110              a2faa651          3/ 0 txs/mgas         13/25 peers
2017-12-30 22:12:39 Import    #5114113 (+2)         efe9500b          0/ 0 txs/mgas         12/25 peers
```

and make sure that txs/mgas for multiple blocks is sum over these blocks

**Comment by whilei** [comment]
@splix Just pushed a few changes addressing your notes above. 

screen shot 2017-12-30 at 16 04 57

**Pull request merged**